core/src/calleeFunctions/index.ts

```diff
@@ -23,27 +23,104 @@ const allCalleeFunctions: Record<CalleeNames, CalleeFunctions> = {
 export const getCalleeData = async function (
     network: string,
     collateralType: string,
+    marketId: string,
     profitAddress: string
 ): Promise<string> {
     const collateral = getCollateralConfigByType(collateralType);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
+    const marketData = collateral.exchanges[marketId];
+    if (!marketData || !marketData.callee || !allCalleeFunctions[marketData.callee]) {
         throw new Error(`Unsupported collateral type "${collateralType}"`);
     }
-    return await calleeFuctions.getCalleeData(network, collateral, profitAddress);
+    return await allCalleeFunctions[marketData.callee].getCalleeData(network, collateral, marketId, profitAddress);
 };
 
-const _getMarketPrice = async function (
+export const getMarketDataById = async function (
+    network: string,
+    collateral: CollateralConfig,
+    marketId: string,
+    amount: BigNumber = new BigNumber('1')
+): Promise<MarketData> {
+    const marketData = collateral.exchanges[marketId];
+    if (!allCalleeFunctions[marketData?.callee]) {
+        throw new Error(`Unsupported callee "${marketData?.callee}" for collateral symbol "${collateral.symbol}"`);
+    }
+    let marketUnitPrice: BigNumber;
+    try {
+        marketUnitPrice = await allCalleeFunctions[marketData.callee].getMarketPrice(
+            network,
+            collateral,
+            marketId,
+            amount
+        );
+    } catch (error: any) {
+        const errorMessage = error;
+        marketUnitPrice = new BigNumber(NaN);
+        return {
+            ...marketData,
+            marketUnitPrice,
+            errorMessage,
+        };
+    }
+    return {
+        ...marketData,
+        marketUnitPrice: marketUnitPrice ? marketUnitPrice : new BigNumber(NaN),
+    };
```

Comment:
Why do you need to try catch it here if you already catching it in the parent function `getMarketData` (within the loop)?